### PR TITLE
updated to use puppet.runtime[:http] to remove deprecated function, a…

### PR DESCRIPTION
**Pull request by vamsinm**
updated to use puppet.runtime[:http] to remove deprecated function, added namespace support, added function to access particular key, added option to use combination of cert and approle auth methods for more security

### CODE REVIEW

**Comment by root-expert** [comment]
Hello @vamsinm and thanks for the PR, this module needed some love for sure.

To make it easier for us to review but also for users to be able to understand what was changed, will you be able to split this PR into multiple for each feature/bug?

**Comment by vamsinm** [comment]
> Hello @root-expert, I am very new to ruby and contributing to open source. i could use some help or guidance if anyone willing to to make this faster. i updated it based on requirements and best practices that an organization would need and we are currently using it in test with out any problem. i am struggling with writomg spec for this. can i get any help pease? i will take a look at how to split the PR.

**Comment by root-expert** [comment]
Hello, for some guidance for unit/acceptance tests you can find us on [Libera.chat IRC](https://libera.chat/) (channel #voxpupuli) or over at [PuppetCommuinity Slack](https://puppetcommunity.slack.com/) (same channel) :smile:

**Comment by root-expert** [comment]
I wrote the first unit test, you can use this as a base:

```ruby
it 'raises a Puppet error when auth fails' do
  response = Puppet::HTTP::Response.new(URI('https://vault.doesnotexist:8200/v1/auth/cert/login'), 403, auth_failure_data)
  allow(response).to receive(:body).and_return(auth_failure_data)

  allow(Puppet.runtime[:http]).to receive(:address).and_return('vault.doesnotexist')
  expect(Puppet.runtime[:http]).to receive(:post).with(
    URI('https://vault.doesnotexist:8200/v1/auth/cert/login'),
    '',
    hash_including(
      headers: hash_including('Content-Type' => 'application/json', ),
      options: hash_including('include_system_store' => true)
    )
  ).and_return(response)

  expect {
    function.execute('thepath', 'https://vault.doesnotexist:8200')
  }.to raise_error(Puppet::Error, %r{Received 403 response code from vault at vault.doesnotexist for authentication \(api errors: \["invalid certificate or no client certificate supplied"\]})
end
```

File: lib/puppet/functions/vault_lookup/lookup.rb [hunk 1]
```diff
@@ -2,60 +2,238 @@
...
+    secret_response = if vault_namespace.nil? || vault_namespace == ''
+                        connection.get(
+                          URI(vault_url + '/v1/' + path.to_s),
+                          options: { 'include_system_store' => true },
+                          headers: { 'X-Vault-Token' => vault_token },
+                        )
+                      else
+                        connection.get(
+                          URI(vault_url + '/v1/' + path.to_s),
+                          options: { 'include_system_store' => true },
```
Comment (joshcooper) [inline lib/puppet/functions/vault_lookup/lookup.rb:101]: Same here:

```suggestion
                          options: { include_system_store: true },
```

File: lib/puppet/functions/vault_lookup/lookup.rb [hunk 2]
```diff
@@ -2,60 +2,238 @@
...
+                else
+                  "{\"name\": \"#{vault_cert_role}\"}"
+                end
+
+    response = if vault_namespace.nil? || vault_namespace == ''
+                 connection.post(
+                   URI(vault_url + '/v1/auth/' + vault_cert_path.to_s + '/login'),
+                   role_data,
+                   headers: { 'Content-Type' => 'application/json' },
+                   options: { 'include_system_store' => true },
```
Comment (joshcooper) [inline lib/puppet/functions/vault_lookup/lookup.rb:143]: Same here:

```suggestion
                   options: { include_system_store: => true },
```

Comment (vamsinm) [reply lib/puppet/functions/vault_lookup/lookup.rb:143]: @joshcooper when changed to `options: { include_system_store: => true }`  works fine with `options: { 'include_system_store' => true },`
```
Error: Failed to apply catalog: Received 400 response code from vault at https://vault.host:8200 for authentication (api errors: ["1 error occurred:\n\t* error running lookahead function for mfa: no client certificate found\n\n"])
```

Comment (joshcooper) [reply lib/puppet/functions/vault_lookup/lookup.rb:143]: Using a symbol is correct. The client cert issue is https://tickets.puppetlabs.com/browse/PUP-11522.

**Comment by bastelfreak** [comment]
Hi,
I think this PR is obsolete since #50 got merged? If you think we still should merge this, please reopen and rebase it.

**Comment by binford2k** [comment]
@vamsinm now that #50 has been merged, can you rework this into a couple PRs against HEAD for the namespacing, etc features?

**Pull request closed without merging**